File: 74. Search_a_2D_Matrix/solution.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        for(int i = 0 ; i < matrix.size() ; i++){
            if(matrix[i][0] == target)
                return true;
            
            if(i + 1 < matrix.size() && matrix[i + 1][0] < target)
                continue;
            
            for(int j = 0 ; j < matrix[0].size() ; j++){
                if(matrix[i][j] == target)
                    return true;
                if(j + 1 < matrix[0].size() && matrix[i][j] < target && target < matrix[i][j+1])
                    return false;
            }
        }
        return false;
        
    }
};

File: 74. Search_a_2D_Matrix/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int left = 0;
        int right = matrix.length * matrix[0].length - 1;
        
        while(left <= right){
            
            int mid = (left + right) / 2;
            
            int mid_num = matrix[mid / matrix[0].length][mid % matrix[0].length];
            
            if(mid_num == target)
                return true;
            else if(mid_num < target)
                left = mid + 1;
            else
                right = mid - 1;
            
        }
        
        return false;
        
    }
}